privacyidea/lib/tokens/radiustoken.py

```diff
@@ -143,15 +144,193 @@ def update(self, param):
+
+        log.debug("***[radius token: create challenge]")
+        log.debug("***loaded challenge {0}".format(state))
+
+        db_challenge = Challenge(self.token.serial,
+                                 transaction_id=transactionid,
+                                 data=state,
+                                 challenge=message,
+                                 validitytime=validity)
+        db_challenge.save()
+        self.challenge_janitor()
+        return True, message, db_challenge.transaction_id, attributes
+
+    def is_challenge_response(self, passw, user=None, options=None):
+        """
+        This method checks, if this is a request, that is the response to
+        a previously sent challenge.
+
+        Since this is RADIUS, we need to additionally check to see if the
+        current request is going to spawn another challenge
+
+        :param passw: password, which might be pin or pin+otp
+        :type passw: string
+        :param user: the requesting user
+        :type user: User object
+        :param options: dictionary of additional request parameters
+        :type options: dict
+        :return: true or false
+        :rtype: bool
+        """
+        options = options or {}
+        otpval = passw
+        challenge_response = False
+
+        log.debug("***[radius token: is challenge response]")
+
+        # clear the radius_result since this is the first function called in the chain
+        # this value will be utilized to ensure we do not check_radius more than once in the loop
+        options.update({'radius_result': None})
+
+        # fetch the transaction_id
+        transaction_id = options.get('transaction_id')
+        if transaction_id is None:
+            transaction_id = options.get('state')
+
+        if transaction_id:
+            # get the challenges for this transaction ID
+            challengeobject_list = get_challenges(serial=self.token.serial,
+                                                  transaction_id=transaction_id)
+
+            for challengeobject in challengeobject_list:
+                if challengeobject.is_valid():
+                    # challenge is still valid, pull the state and message and other data
+                    state = challengeobject.data
+
+                    # we need to check against radius before continuing
+                    # this causes a bit of an issue since we will know
+                    # if the challenge is successful before we try
+                    log.debug("***[radius token: challenge response state check] {0}".format(state))
+                    res = self.check_radius(otpval, options=options, radius_state=state)
+                    if res != 1:  # we are not getting challenged again so we can stop the chain
+                        challenge_response = True
+                    else:
+                        challengeobject.delete()  # a new challenge will be created
```

Comment:
There might be a new challenge in the RADIUS server we are contacting, but to me it currently looks like as if there is no new challenge in privacyIDEA!?

Reply:
I think this was just a misleading comment.